### PR TITLE
Py3k: Prefer Py3k over Py2k if available

### PR DESCRIPTION
**Pull request by jdemel**
This is a simple change that pushes CMake to search for Python3 first.
Only if we can't find a Py3k install, we move on to search for Py2k.

This PR deals with #195

### CODE REVIEW

**Review by michaelld — CHANGES_REQUESTED** [review]
I like this change. We do need to keep Py2 around at least for Volk 2, IMHO. As such, would you add some logic to determine which `PythonInterp` was found (2 or 3), and then look just for the `PYTHON_EXECUTABLE` of that version? Thus, for example, of `PythonInterp` found is version 2, then don't look for the Py3 `PYTHON_EXECUTABLE`. Should be straight forward, I think, and not a huge change.

**Comment by jdemel** [comment]
Basically, we want 2 cmake variables to be set
1. `PYTHONINTERP_FOUND` to indicate we can use Python
2. `PYTHON_EXECUTABLE` to use this specific Python version.

Previously, the code worked like this:
If `PYTHON_EXECUTABLE` is already set, e.g. by user, we just use it and `set(PYTHONINTERP_FOUND TRUE)`. This is still the same.

If we did not set `PYTHON_EXECUTABLE` we try to find it. This is where I my PR takes effect.
With `find_package(PythonInterp 3)` we search for a Py3k. This will set both cmake variables. No further action required.
Only if we can't find Py3k, we go on and search for Py2k.

Finally, if that fails as well, we try:
`find_program(PYTHON_EXECUTABLE NAMES python3 python python2 python2.7)`
This should search for executable names in order of appearance. Thus, we search for `python3` first.

Since we require Python 2.7+, we can remove 
```
#make sure we can use -B with python (introduced in 2.6)
if(PYTHON_EXECUTABLE)
    execute_process(
        COMMAND ${PYTHON_EXECUTABLE} -B -c ""
        OUTPUT_QUIET ERROR_QUIET
        RESULT_VARIABLE PYTHON_HAS_DASH_B_RESULT
    )
    if(PYTHON_HAS_DASH_B_RESULT EQUAL 0)
        set(PYTHON_DASH_B "-B")
    endif()
endif(PYTHON_EXECUTABLE)
```
in `cmake/Modules/VolkPython.cmake

I updated `.travis.yml` as well I hope the CI system does work again with my patch.

**Review by michaelld — COMMENTED** [review]
Gotcha. Yes, that all makes sense. Good changes! Let me test locally with different Python 2 and 3 & verify before finalizing.

File: cmake/Modules/VolkPython.cmake
Comment (michaelld) [inline cmake/Modules/VolkPython.cmake:52]: according to CMake 3.16.0's `FindPythonInterp.cmake`:
```
deprecated:: 3.12
  Use :module:`FindPython3`, :module:`FindPython2` or :module:`FindPython` instead.
```
Looks like the new usage is `find_package(Python3 COMPONENTS Interpreter)`

Same for the 2.7 call.

Comment (michaelld) [reply cmake/Modules/VolkPython.cmake:52]: Looks like instead of `` we now do `set(Python3_FIND_STRATEGY VERSION)` ... which is the default so we  can just remove the `Python_ADDITIONAL_VERSIONS` I think.
```
``Python3_FIND_STRATEGY``
  This variable defines how lookup will be done.
  The ``Python3_FIND_STRATEGY`` variable can be set to empty or one of the
  following:

  * ``VERSION``: Try to find the most recent version in all specified
    locations.
    This is the default if policy :policy:`CMP0094` is undefined or set to
    ``OLD``.
  * ``LOCATION``: Stops lookup as soon as a version satisfying version
    constraints is founded.
    This is the default if policy :policy:`CMP0094` is set to ``NEW``.
```

File: cmake/Modules/VolkPython.cmake
Comment (michaelld) [inline cmake/Modules/VolkPython.cmake:54]: This would be `Python3_Interpreter_FOUND`.

File: cmake/Modules/VolkPython.cmake
Comment (michaelld) [inline cmake/Modules/VolkPython.cmake:55]: Thinking somewhere in here you'd want logic setting a common `PYTHON_INTERPRETER_FOUND` and `PYTHON_EXECUTABLE` ... whichever of Py3 or Py2 was found .... variables that will be common once set regardless of which Py is found and how (old / deprecated or current new, or some future way).

**Review by michaelld — CHANGES_REQUESTED** [review]
A few changes will bring the cmake up to date, but it's a little more complex than I'd hoped.

File: cmake/Modules/VolkPython.cmake
Comment (jdemel) [reply cmake/Modules/VolkPython.cmake:52]: Yeah, the new, in fact too new, way would be `find_package(Python COMPONENTS Interpreter)` [cf here](https://cmake.org/cmake/help/v3.16/module/FindPython.html#module:FindPython). This call would prefer Py3k over Py2k, thus effectively implement the same behavior I intend to implement.
I browsed through the CMake documentation and it seems this function was only added in CMake 3.12. And we currently require CMake 3.8. 

With CMake 3.12+ we could simply do this:
```
if(PYTHON_EXECUTABLE)
    set(Python_EXECUTABLE ${PYTHON_EXECUTABLE})
else(PYTHON_EXECUTABLE)
    find_package(Python COMPONENTS Interpreter)
endif(PYTHON_EXECUTABLE)

#make the path to the executable appear in the cmake gui
set(PYTHON_EXECUTABLE ${Python_EXECUTABLE} CACHE FILEPATH "python interpreter")
```
Since `grep -irnI PYTHONINTERP_FOUND .` did not reveal any use of this variable outside this section, we don't need to set this variable anymore.

I suggest to stick with my implementation for now. We can update this section after we move to CMake 3.12+ as a minimum. I don't like to make this section version dependent. This would make it more difficult to read.
@michaelld how would you proceed?

Comment (michaelld) [reply cmake/Modules/VolkPython.cmake:52]: OK yeah given the CMake min version, let's leave this code as it is. Would you please add a note about changing up the code when CMake >= 3.12? That way we're leaving something useful behind!

**Review by michaelld — CHANGES_REQUESTED** [review]
LGTM after that comment addition.

**Comment by jdemel** [comment]
I added a comment `FUTURE TODO` with a short hint on how to handle this with CMake 3.12+. I hope it'll help us then. Since Ubuntu 18.04 ships with CMake 3.11, I doubt we'll move to a more recent CMake version any time soon.

**Review by michaelld — APPROVED** [review]
Thanks for the comment. LGTM.

**Comment by michaelld** [comment]
CI passes that we care about; failures are not due to this PR. Maintainer approves. Merging.

**Pull request merged**
